packages/connect-web/src/popup/index.ts

```diff
@@ -159,65 +241,67 @@ export class PopupManager extends EventEmitter {
-        if (!this.extensionPort) return;
-        const port = this.extensionPort;
-        const { data } = message;
-        if (!data || typeof data !== 'object') return;
-
-        if (data.type === POPUP.ERROR) {
-            // handle popup error
-            const errorMessage =
-                data.payload && typeof data.payload.error === 'string' ? data.payload.error : null;
-            this.emit(POPUP.CLOSED, errorMessage ? `Popup error: ${errorMessage}` : null);
-            this.clear();
-        } else if (data.type === POPUP.LOADED) {
-            if (this.popupPromise) {
-                this.popupPromise.resolve();
+    private injectContentScript = (tabId: number) => {
+        chrome.permissions.getAll(permissions => {
+            if (permissions.permissions?.includes('scripting')) {
+                chrome.scripting
+                    .executeScript({
+                        target: { tabId },
+                        // content script is injected into body of func in build time.
+                        func: () => {
+                            // <!--content-script-->
+                        },
+                    })
+                    .then(() => this.logger.debug('content script injected'))
+                    .catch(error => this.logger.error('content script injection error', error));
+            } else {
+                // When permissions for `scripting` are not provided 3rd party integrations have include content-script.js manually.
             }
-            this.iframeHandshake.promise.then(payload => {
-                port.postMessage({
-                    type: POPUP.INIT,
-                    payload: {
-                        ...payload,
-                        settings: this.settings,
-                    },
-                });
+        });
+    };
+
+    handleCoreMessage(message: Message<CoreEventMessage>) {
+        if (message.type === POPUP.CORE_LOADED) {
+            this.channel.postMessage({
+                type: POPUP.HANDSHAKE,
+                // in this case, settings will be validated in popup
+                payload: { settings: this.settings },
+            });
+            this.handshakePromise?.resolve();
+        }
+        if (message.type === POPUP.CLOSED) {
+            // When popup is closed we should create a not-real response as if the request was interrupted.
+            // Because when popup closes and TrezorConnect is living there it cannot respond, but we know
+            // it was interrupted so we safely fake it.
+            this.channel.resolveMessagePromises({
+                code: 'Method_Interrupted',
+                error: POPUP.CLOSED,
             });
+        }
+    }
+
+    handleExtensionMessage(data: Message<CoreEventMessage>) {
+        if (
+            data.type === POPUP.ERROR ||
+            data.type === POPUP.LOADED ||
+            data.type === POPUP.BOOTSTRAP
+        ) {
+            this.handleMessage(data);
```

Comment:
I think this would be more readable and less error-prone, but just an opinion.
```suggestion
        if (
            [POPUP.ERROR, POPUP.LOADED, POPUP.BOOTSTRAP].includes(data.type)
        ) {
            this.handleMessage(data);
```

Reply:
Sure, seems reasonable

Reply:
Tried it, but had to revert it, TS didn't like it